### PR TITLE
Aerospike mvp

### PR DESCRIPTION
**Pull request by geobeau**

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+// Start the probing scheduler:
+// - listen for topology changes
+// - start and stop probes
+func (ps *ProbingScheduler) Start() {
+	for {
+		newTopology := <-ps.topologyUpdateChan
+
+		level.Info(ps.logger).Log("msg", "New topology received, updating...")
+
+		toStopendpoints, toAddEndpoints := ps.currentTopology.Diff(&newTopology)
```
Comment (ashangit) [inline pkg/scheduler/scheduler.go:100]: toStopEndpoints

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+		if !check_performed {
+			break
+		}
+	}
+}
+
+func (pw *ProberWorker) StartProbing() {
+	level.Info(pw.logger).Log("msg", "starting probing")
+
+	if len(pw.checks) < 1 {
```
Comment (ashangit) [inline pkg/scheduler/scheduler.go:178]: we probably need to check this before to start it in startNewWorker so it avoid creating chan and add it in workerControlChans...

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		return
+	}
+
+	lastChecks := make([]time.Time, len(pw.checks))
+	shortestInterval := pw.checks[0].Interval
+
+	err := pw.endpoint.Connect()
+	if err != nil {
+		level.Error(pw.logger).Log("msg", "Probe failure during connection", "err", err)
+		return
```
Comment (ashangit) [inline pkg/scheduler/scheduler.go:189]: perhaps we could init the probing in blocking way in startNewWorker so we can trap init issues and then run the probing part in a goroutine

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	}
+
+	for i, check := range pw.checks {
+		lastChecks[i] = time.Now()
+		check.PrepareFn(pw.endpoint)
+		if shortestInterval > check.Interval {
```
Comment (ashangit) [inline pkg/scheduler/scheduler.go:195]: plz add a comment here to explain why we look for shortestInterval
-> shortestInterval will be used as ticker for all checks. findWork will then checks if we wait enough time before to run each checks. Means that we can have an approximate on duration between checks

File: probes/aerospike/config.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	TLSTag string `yaml:"tls_tag,omitempty"`
+	// Metadata key to get the Hostname to use for TLS auth (only used if tlsTag is set)
+	TLSHostnameMetaKey string `yaml:"tls_hostname_meta_key,omitempty"`
+}
+
+var (
+	defaultAerospikeClient = AerospikeEndpointConfig{
+		AuthEnabled:        true,
+		AuthExternal:       true,
+		UsernameEnv:        "AEROSPIKE_USERNAME",
```
Comment (ashangit) [inline probes/aerospike/config.go:45]: Why not getting both of them from config file?

**Review by ashangit — CHANGES_REQUESTED** [review]
Overall looks fine
I have little concern with each check not running in dedicated goroutine and not exactly running at each define interval.
Also would be nice to have a timeout per checks

File: README.md [hunk 6]
```diff
@@ -1 +1,126 @@
...
+Useful to calculate SLIs about a single node (such as latency)
+
+- **Cluster**: executed every interval per cluster.
+Useful to calculate SLIs about a whole cluster (such as durability)
+
+# Usage
+
+### Building
+```
+go build -o build/aerospike_probe probes/aerospike/*.go
```
[the remark this answers is not in the document]
Comment (geobeau) [reply README.md]: Done

File: config.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+discovery:
```
[the remark this answers is not in the document]
Comment (geobeau) [reply config.yaml]: Done

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,10 @@
+.PHONY: test build build_linux
+
+test:
```
Comment (ashangit) [inline Makefile:3]: can we also add run of go vet and go fmt

.PHONY: fmt
fmt:
	@out=$$(gofmt -d -e -s ${GOPKGS_DIRS} 2>&1) && [ -z "$${out}" ]  || (echo "$${out}";exit 1)

.PHONY: update-fmt
update-fmt:
	@gofmt -l -w ${GOPKGS_DIRS}

.PHONY: vet
vet:
	@go vet ${GOPKGS}

test: fmt vet
 go test...

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+		go test ./...
+
```
Comment (ashangit) [inline Makefile:5]: Can we rely on vendor mechanism to get all dependencies locally

File: probes/aerospike/config.go (hunk 5 above)
Comment (ashangit) [reply probes/aerospike/config.go:45]: env variables used by a process can be read by doing cat /proc/pid/environ
so I'm not sure that pattern add any security

**Review by ashangit — CHANGES_REQUESTED** [review]

File: Makefile (hunk 8 above)
Comment (geobeau) [reply Makefile:3]: Done

File: Makefile (hunk 9 above)
Comment (geobeau) [reply Makefile:5]: Done

File: probes/aerospike/config.go (hunk 5 above)
Comment (geobeau) [reply probes/aerospike/config.go:45]: /proc/ is in memory if I'm not mistaken.
Using env also prevent committing credentials by mistake. It is also simple to inject in both Mesos and Kubernetes (directly from a secret).

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	"crypto/rand"
+	"encoding/hex"
+)
+
+var (
+	MetricSuffix           = "blackbox_prober"
+	MetricHistogramBuckets = []float64{.001, .0025, .005, .010, .015, .020, .025, .030, .040, .050, .060, .075, .100, .250, .500, 1, 2.5, 5, 10, 15, 30, 45, 60}
+)
+
+func Contains(s []string, e string) bool {
```
[the remark this answers is not in the document]
Comment (geobeau) [reply pkg/utils/utils.go:13]: Done

File: probes/aerospike/config.go (hunk 5 above)
Comment (ashangit) [reply probes/aerospike/config.go:45]: /proc/ is in memory if I'm not mistaken -> yes but will be there any time the process run which mean always on k8s so I wanted to point that this doesn't provide any security

File: Makefile (hunk 9 above)
Comment (ashangit) [reply Makefile:5]: As discussed will be done in next commit

File: pkg/common/config.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	HttpListenAddr string         `yaml:"http_listen_addr,omitempty"`
+	ConfigPath     string         `yaml:"config_path,omitempty"`
+}
+
+func AddFlags(a *kingpin.Application, cfg *ProbeConfig) {
+	a.HelpFlag.Short('h')
+	a.Flag("web.listen-address", "Address to listen on for UI, API, and telemetry.").
+		Default("0.0.0.0:8080").StringVar(&cfg.HttpListenAddr)
+	a.Flag("config.path", "Path to the probe configuration file").
+		Default("config.yaml").StringVar(&cfg.ConfigPath)
```
Comment (jfwm2) [inline pkg/common/config.go:30]: IMO you should either have a config.yaml file or change the default to an existing one

File: config_aerospike.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    services:
+    - aerospike
+client_config:
+  ### AUTH ### 
+  # Enable authentication
+  auth_enabled: true
+  # Use external authentication (https://docs.aerospike.com/docs/guide/security/access-control.html#external-authentication-with-ldap)
+  auth_external: true
+  # Env variable to use to retrieve the credentials
+  username_env: "AEROSPIKE_USERNAME"
```
Comment (jfwm2) [inline config_aerospike.yaml:19]: This should be mentioned in the README file

File: pkg/scheduler/scheduler.go [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+			level.Debug(pw.logger).Log("msg", "Probe endpoint refreshed")
+			err := pw.endpoint.Refresh()
+			if err != nil {
+				level.Error(pw.logger).Log("msg", "Error while refreshing", "err", err)
+				EndpointFailureTotal.WithLabelValues("refresh", pw.endpoint.GetName())
+			} else {
+				EndpointSuccessTotal.WithLabelValues("refresh", pw.endpoint.GetName())
+			}
+		case <-checkTicker.C:
+			fmt.Println("Called")
```
Comment (jfwm2) [inline pkg/scheduler/scheduler.go:251]: Message is too broad

**Review by jfwm2 — APPROVED** [review]
LGTM (a few minor comments added)

**Pull request merged**